client_plugin/drivers/shared/dockerops/dockerops.go

```diff
@@ -142,13 +175,165 @@ func (d *DockerOps) VolumeRemove(volName string) error {
+	// ResolutionModeVIP indicates that networking uses virtual IPs.
+	service.EndpointSpec = &swarm.EndpointSpec{
+		Mode:  swarm.ResolutionModeVIP,
+		Ports: exposedPorts,
+	}
+
+	//Start the service
+	resp, err := d.Dockerd.ServiceCreate(context.Background(),
+		service, options)
+	if err != nil {
+		log.Warningf("Failed to create file server for volume %s. Reason: %v",
+			volName, err)
+		return 0, "", false
+	}
+
+	// Wait till service container starts
+	ticker := time.NewTicker(checkDuration)
+	defer ticker.Stop()
+	timer := time.NewTimer(sambaRequestTimeout)
+	defer timer.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			log.Infof("Checking status of file server container...")
+			port, isRunning := d.isFileServiceRunning(resp.ID, volName)
+			if isRunning {
+				return int(port), volName + serviceNamePostfix, isRunning
+			}
+		case <-timer.C:
+			log.Warningf("Timeout reached while waiting for file server container for volume %s",
+				volName)
+			return 0, "", false
+		}
+	}
+}
+
+// isFileServiceRunning - Checks if a file service container is running
+// It takes some time from service being up to a container running.
+// This function returns true only if the container is up.
+func (d *DockerOps) isFileServiceRunning(servID string, volName string) (uint32, bool) {
+	var port uint32
+	// Grep the samba service running for this volume using service ID
+	serviceFilters := filters.NewArgs()
+	serviceFilters.Add("id", servID)
+	services, err := d.Dockerd.ServiceList(context.Background(),
+		dockerTypes.ServiceListOptions{Filter: serviceFilters})
+	if err != nil {
+		log.Warningf("Failed to check if file server for volume %s was started. %v", volName, err)
+		return port, false
+	}
+	if len(services) < 1 {
+		log.Warningf("No service returned for volume %s. Service not started properly.", volName)
+		return port, false
+	}
+
+	port = services[0].Endpoint.Ports[0].PublishedPort
+
+	// Grep all tasks for the service returned and verify that their state is running
+	taskFilter := filters.NewArgs()
+	for _, service := range services {
+		taskFilter.Add("service", service.ID)
+	}
+	tasks, err := d.Dockerd.TaskList(context.Background(),
+		dockerTypes.TaskListOptions{Filter: taskFilter})
+	if err != nil {
+		log.Warningf("Failed to get task list for file service for volume %s. %v", volName, err)
+		return port, false
+	}
+	for _, task := range tasks {
+		if task.Status.State != swarm.TaskStateRunning {
+			log.Infof("File server not running for volume %s", volName)
+			return port, false
+		}
+	}
+	return port, true
 }
 
 // StopSMBServer - Stop SMB server
-func (d *DockerOps) StopSMBServer(volName string) bool {
+func (d *DockerOps) StopSMBServer(volName string) (int, string, bool) {
 	log.Errorf("stopSMBServer to be implemented")
+	return 0, "", true
+}
+
+// loadFileServerImage - Load the file server image present
+// in the plugin to Docker images
+func (d *DockerOps) LoadFileServerImage() bool {
+	file, err := os.Open(fileServerPath)
```

Comment:
From what I could gather from the Go Doc, ImageLoad() takes an io.Reader. This is what I got from Open().

ImageLoad() returns an io.ReadCloser() which I called at line 333: err = resp.Body.Close(). 

Is this not sufficient to close the file? Conventionally after an open() a close() should be called on the same object. But then what is the io.ReadCloser() for?

https://godoc.org/github.com/moby/moby/client#Client.ImageLoad